django/db/models/fields/related_descriptors.py

```diff
@@ -67,9 +67,17 @@ class Child(Model):
 from django.db import connections, router, transaction
 from django.db.models import Q, signals
 from django.db.models.query import QuerySet
+from django.db.models.query_utils import DeferredAttribute
 from django.utils.functional import cached_property
 
 
+class ForeignKeyDeferredAttribute(DeferredAttribute):
+    def __set__(self, instance, value):
+        if self.field.is_cached(instance):
+            self.field.delete_cached_value(instance)
```

Comment:
I have one concern :disappointed: Should we clear cached instance when a value is the same? because this leads to unnecessary queries. Maybe we should check value before clearing an instance, e.g.
```diff
diff --git a/django/db/models/fields/related_descriptors.py b/django/db/models/fields/related_descriptors.py
index 346f67289e..0b51b68ebe 100644
--- a/django/db/models/fields/related_descriptors.py
+++ b/django/db/models/fields/related_descriptors.py
@@ -73,7 +73,7 @@ from django.utils.functional import cached_property
 
 class ForeignKeyDeferredAttribute(DeferredAttribute):
     def __set__(self, instance, value):
-        if self.field.is_cached(instance):
+        if instance.__dict__.get(self.field.attname) != value and self.field.is_cached(instance):
             self.field.delete_cached_value(instance)
         instance.__dict__[self.field.attname] = value
```

You can find below tests that don't work anymore after the second commit.

```diff
diff --git a/tests/many_to_one/tests.py b/tests/many_to_one/tests.py
index 5efe137730..cf3bea86d4 100644
--- a/tests/many_to_one/tests.py
+++ b/tests/many_to_one/tests.py
@@ -164,6 +164,8 @@ class ManyToOneTests(TestCase):
         self.assertFalse(Parent.bestchild.is_cached(parent))
         self.assertEqual(parent.bestchild, child2)
         self.assertTrue(Parent.bestchild.is_cached(parent))
+        parent.bestchild_id = child2.pk
+        self.assertTrue(Parent.bestchild.is_cached(parent))
 
     def test_selects(self):
         self.r.article_set.create(headline="John's second story", pub_date=datetime.date(2005, 7, 29))
diff --git a/tests/one_to_one/tests.py b/tests/one_to_one/tests.py
index 76f9d75417..0bc94bbb32 100644
--- a/tests/one_to_one/tests.py
+++ b/tests/one_to_one/tests.py
@@ -213,6 +213,8 @@ class OneToOneTests(TestCase):
         self.assertFalse(UndergroundBar.place.is_cached(b))
         self.assertEqual(b.place, self.p2)
         self.assertTrue(UndergroundBar.place.is_cached(b))
+        b.place_id = self.p2.pk
+        self.assertTrue(UndergroundBar.place.is_cached(b))
 
     def test_related_object_cache(self):
         """ Regression test for #6886 (the related-object cache) """
```